source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+    return copy;
+  }
+
+  /**
+     Function body. Initial value: `;`.
+   */
+
+  string body_;
+
+  /**
+     Return a new Function object with the body_ attribute set to value.
+  */
+
+  Function setBody(string value)
+  {
+    mixin replaceAttribute!"body_";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure int answer();
+    mixin(
+      refract!(answer, "answer").setName("theAnswer")
+      .setBody("{ return 42; }")
+      .mixture);
+    static assert(theAnswer() == 42);
+  }
+
+  /**
+     Function attributes.
+     Initial value: `__traits(getAttributes, fun)`.
+   */
+
+  ulong attributes;
+
+  /**
+     Set `attributes`. Return `this`.
+   */
+
+  Function setAttributes(uint value)
+  {
+    mixin replaceAttribute!"attributes";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    nothrow int answer();
+    enum model = refract!(answer, "answer");
+    with (FunctionAttribute)
+    {
+      mixin(
+        model
+        .setName("pureAnswer")
+        .setAttributes(model.attributes | pure_)
+        .mixture);
+      static assert(functionAttributes!pureAnswer & pure_);
+      static assert(functionAttributes!pureAnswer & nothrow_);
+    }
+  }
+
+  /**
+     If `true`, prefix generated function with `static`. Initial value: `true`
+     if the refracted function is a static *member* function.
+   */
+
+  bool static_;
+
+  /**
+     Return a new Function object with the static_ attribute set to value.
+   */
+
+  Function setStatic(bool value)
+  {
+    mixin replaceAttribute!"static_";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    struct Question
+    {
+      static int answer() { return 42; }
+    }
+    mixin(
+      refract!(Question.answer, "Question.answer")
+      .setStatic(false)
+      .setBody("{ return Question.answer; }")
+      .mixture);
+    static assert(answer() == 42);
+  }
+
+  /**
+     User defined attributes.
+     Initial value:
+     `bolts.experimental.refraction.ParameterAttribute!(fun, parameterIndex..., attributeIndex...)`.
+   */
+
+  string[] udas;
+
+  /**
+     Return a new Function object with the udas attribute set to value.
+   */
+
+  Function setUdas(immutable(string)[] value)
+  {
+    mixin replaceAttribute!"udas";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    import std.typecons : tuple;
+    @(666) int answer();
+
+    mixin(
+      refract!(answer, "answer")
+      .setName("answerIs42")
+      .setUdas(["@(42)"])
+      .mixture);
+    static assert(__traits(getAttributes, answerIs42).length == 1);
+    static assert(__traits(getAttributes, answerIs42)[0] == 42);
+  }
+
+  /**
+     Return a string representing the entire function definition or declaration.
+   */
+
+  string mixture()
+  {
+    return join(
+      udas ~
+      attributeMixtureArray() ~
+      [
+        returnType,
+        name ~ "(" ~ parameterListMixtureArray.join(", ") ~ ")",
+      ], " ") ~
+      body_;
+  }
+
+  string[] parameterListMixtureArray()
```

Comment:
Do all these other mixtures have to be public?

Reply:
> Do all these other mixtures have to be public?

Yes. It makes it possible to add parameters, combine parameter lists, etc.

Reply:
Are they used in any of your usecases outside the library? They seem to be mostly one-liners (except for the attributes one).

Reply:
> Are they used in any of your usecases outside the library? They seem to be mostly one-liners (except for the attributes one).

[In openmethods](https://github.com/jll63/openmethods.d/blob/master/source/openmethods.d#L472) I need to edit the parameter themselves. I use `argumentMixture` in openmethods and `argumentMixtureArray` in my now obsolete ducktyping project. In the latter case, I need the array variant because I need to add arguments. As for `parameterListMixtureArray`, I think that omitting it would make the API incomplete and irregular. It's only coincidental that it is also used to implement another public function.

Reply:
I think functions that are obvious and can be implemented as a one liner generally should be part of a public API except for in 2 situations: 1) if they are used a lot and 2) if they provide some kind of portability (across compilers, platforms, etc). I'm not sure in this case though. 

In this case the only one that I think is actually non-trivial is `attributeMixtureArray`, but adding that and not the others would make the API, as you say, irregular. So I propose a property function named `attributesAsStrings` or `attributeStrings` that return the attributes as `string[]` and cut out the others.

Additionally, it feels a bit off that a mixture is not mixable. I'd think that anything with mixture at the end can be used as a mixin, but I don't think you can mixin argumentMixture anywhere no?

Reply:
> In this case the only one that I think is actually non-trivial is `attributeMixtureArray`, but adding that and not the others would make the API, as you say, irregular. So I propose a property function named `attributesAsStrings` or `attributeStrings` that return the attributes as `string[]` and cut out the others.

Ah, I disagree...you see, I didn't dream up this API out of context. When I set out to support function attributes and parameter storage classes in openmethods (in response to https://github.com/jll63/openmethods.d/issues/13 - note that it stayed opened for 2 years), the code that generates functions went through many iterations. At one point *all* the generated code was created by string mixins, themselves created via `format`. Essentially all the compile-time part of openmethods had turned into an ugly bunch of C-style macros. Unfortunately, I was not aware of Adam's Tip of teh Week back then.

I started experimenting with string interpolation, and that helped a little. Then I came up with a `FunctionBuilder`. It helped more, but still, what emerged was that sometimes I needed the arguments as a string, and sometimes as an array (the latter, typically, when I needed to add an argument at the front of the list). The code was better, but still difficult to read because it juggled strings around pretty much like the `format` version.

So I added the helpers (again it went though many iterations, some of them visible in openmenthod's master branch) with the goal of making the generation code as readable as possible. At that point `FuinctionBuilder` had become a sort of DSL (and bear in mind that the goal is not to replicate functions, but to transform them easily).

At that point these was a possibility that `FunctionBuilder` was in fact tailored to openmethods and was not generally useful,  but then I implemented my (unpublished, obsolete) ducktyping library, and it appeared that several of these partial mixture, returning either a string or an array, were useful in the second library as well. IOW, if these functions had not been in the (then named) `FunctionBuilder`, I would have had to duplicate them in openmethods and ducktyping.

As a counter-example, this code from ducktyping:

```d
  static auto makePointer(Function fun)
  {
    auto name = fun.name;
    return fun
      .removeMutabilityAttributes
      .insertParameters(
        0, Parameter(
          fun.copyMutabilityAttribute("void*")))
      .setName("function")
      .setBody("")
      .mixture ~ " %s%d;".format(name, fun.index);
  }

  struct Vtbl
  {
    mixin(refract!(Interface, "Interface").map!makePointer.join("\n"));
  }
```

...uses two helper functions:

```d
string copyMutabilityAttribute(Function fun, string type)
{
  with (FunctionAttribute) {
    return fun.attributes & const_ ? "const " ~ type
      : fun.attributes & immutable_ ? "immutable " ~ type
      : type;
  }
}

private auto removeMutabilityAttributes(immutable(Function) fun)
{
  return fun.setAttributes(
    fun.attributes & ~(FunctionAttribute.const_ | FunctionAttribute.immutable_));
}
```

These are clearly ducktyping-specific, and would have no use in openmethods. OTOH, `argumentMixture` is used in both, and I predict that most use cases of `refract` will need it as well.

All that to say, the API is not a case of armchair design. It grew up from considering the converging and diverging needs of two libraries.

> Additionally, it feels a bit off that a mixture is not mixable. I'd think that anything with mixture at the end can be used as a mixin, but I don't think you can mixin argumentMixture anywhere no?

The "mixture" stem also came after many iterations. Initially, they were called `asString` (following your convention in bolts) and `asStringArray`. But then I felt that these strings are created following strict constraints, with one ultimate goal: land inside a string mixin. So I renamed them to `asMixin` and `asMixinArray`. But then those strings were not complete mixins - as you justly point out -, more like fragments of mixins. Thus they became `asFragment` and `asFragmentArray` (this time inspired by [Beta](https://beta.cs.au.dk/Manuals/latest/beta/fragment.html)). But that lost the focus on D's mixins. Eventually I came up with mixture. A mixture is not a mixin, but a bit of mixin that, mixed with other mixtures, eventually builds into a string mixin.

In the near future I hope to popularize Adam's tip, and promote the word "mixture" to mean "bit of source code that, when string-mixed-in, will work correctly even across modules, like Adam initially explained".